[1LP][RFR] Add compare_hosts function to handle navigation and checking compare items

prichard77 opened the pull request:
## Purpose or Intent

- __Enhancement__ for a common function to do the navigation, checking of items, and selecting compare item in dropdown for comparison of hosts tests. This is the first phase of development of common compare functionality for entities. I'll be extending to other entity types (and probably implementing the functionality via class methods etc.) in future PRs. Existing Host compare tests have been updated to use compare_hosts.


### PRT Run

{{ pytest: --long-running --use-provider rhv43 cfme/tests/infrastructure/test_host.py::test_infrastructure_hosts_navigation_after_download_from_compare cfme/tests/infrastructure/test_host.py::test_infrastructure_hosts_compare cfme/tests/infrastructure/test_providers.py::test_compare_templates }}

prichard77 commented:
I am creating this for host entities, but have plans to extend to other entities in the future. I'll need to do more investigation before I determine how I want to extend this, but want to get it working for hosts first.

dajoRH commented:
Lint report for commit 99e9cf3232ad161a6f88ba4282936a493200ad5f:

`cfme/tests/infrastructure/test_host.py`:
- :bangbang: Line 516:1: **E265** *block comment should start with '# '*
- :bangbang: Line 520:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 522:13: **E261** *at least two spaces before inline comment*
- :bangbang: Line 522:14: **E262** *inline comment should start with '# '*
- :bangbang: Line 524:13: **E261** *at least two spaces before inline comment*
- :bangbang: Line 524:14: **E262** *inline comment should start with '# '*
- :red_circle: Line 530:34: **F821** *undefined name 'appliance'*
- :bangbang: Line 530:69: **E261** *at least two spaces before inline comment*
- :bangbang: Line 540:1: **E303** *too many blank lines (4)*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 7db9ce19a21c4d2498a0c65dcd40f259d6050a01:

`cfme/tests/infrastructure/test_host.py`:
- :bangbang: Line 525:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 527:13: **E261** *at least two spaces before inline comment*
- :bangbang: Line 529:13: **E261** *at least two spaces before inline comment*
- :red_circle: Line 535:34: **F821** *undefined name 'appliance'*
- :bangbang: Line 535:69: **E261** *at least two spaces before inline comment*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 40e6696e1d4c3b0fc6e7feb8569c679be65295f7:

`cfme/tests/infrastructure/test_host.py`:
- :bangbang: Line 483:47: **E127** *continuation line over-indented for visual indent*
- :bangbang: Line 492:36: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 495:1: **E265** *block comment should start with '# '*
- :bangbang: Line 532:1: **E265** *block comment should start with '# '*
- :bangbang: Line 535:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 535:86: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 544:66: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 544:68: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 546:18: **E261** *at least two spaces before inline comment*
- :bangbang: Line 568:41: **E231** *missing whitespace after ','*
- :bangbang: Line 571:5: **E265** *block comment should start with '# '*
- :bangbang: Line 578:41: **E231** *missing whitespace after ','*
- :bangbang: Line 578:44: **E261** *at least two spaces before inline comment*
- :bangbang: Line 581:5: **E265** *block comment should start with '# '*
- :bangbang: Line 591:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 594:24: **E261** *at least two spaces before inline comment*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit b8076d06fe0ade406f9c03491d66791243c06152:

`cfme/tests/infrastructure/test_host.py`:
- :bangbang: Line 462:47: **E127** *continuation line over-indented for visual indent*
- :bangbang: Line 471:36: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 474:1: **E265** *block comment should start with '# '*
- :bangbang: Line 511:1: **E265** *block comment should start with '# '*
- :bangbang: Line 514:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 514:86: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 523:66: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 523:68: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 525:18: **E261** *at least two spaces before inline comment*
- :bangbang: Line 547:41: **E231** *missing whitespace after ','*
- :bangbang: Line 550:5: **E265** *block comment should start with '# '*
- :bangbang: Line 557:41: **E231** *missing whitespace after ','*
- :bangbang: Line 557:44: **E261** *at least two spaces before inline comment*
- :bangbang: Line 560:5: **E265** *block comment should start with '# '*
- :bangbang: Line 570:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 573:24: **E261** *at least two spaces before inline comment*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit b7d07305a3c3c5bf65181fc20184ed332298a45c:

`cfme/common/__init__.py`:
- :bangbang: Line 367:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 397:21: **E117** *over-indented*
- :bangbang: Line 403:9: **E265** *block comment should start with '# '*
- :bangbang: Line 404:101: **E501** *line too long (101 > 100 characters)*
- :bangbang: Line 416:1: **E303** *too many blank lines (3)*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 90bdc634871c6ba6a0f9a1a3883e4c59bc6fe417:

`cfme/common/__init__.py`:
- :bangbang: Line 367:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 403:13: **E265** *block comment should start with '# '*
- :bangbang: Line 406:21: **E117** *over-indented*
- :bangbang: Line 418:101: **E501** *line too long (102 > 100 characters)*
- :bangbang: Line 431:1: **E303** *too many blank lines (3)*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 8327c925ad084e7ae5189998d6fa5a3d7a7169d1:

`cfme/common/__init__.py`:
- :red_circle: Line 383:9: **F401** *'cfme.common.host_views.HostsCompareView' imported but unused*
- :red_circle: Line 384:9: **F401** *'cfme.common.host_views.ProviderHostsCompareView' imported but unused*
- :red_circle: Line 385:9: **F401** *'cfme.common.vm.VM' imported but unused*
- :bangbang: Line 392:70: **E261** *at least two spaces before inline comment*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

prichard77 commented:
I had this working for hosts, but it was not done correctly. I added the mixin to collection and when I removed it from Hosts the test does not work. I need to figure this out and make the test work without the mixing Comparable in Hosts. I'd welcome any additional comments.

dajoRH commented:
Lint report for commit 677c63fc135f3888d79cec026189e98618cf92cf:

`cfme/common/__init__.py`:
- :bangbang: Line 389:70: **E261** *at least two spaces before inline comment*

`cfme/tests/infrastructure/test_host.py`:
- :bangbang: Line 491:46: **E261** *at least two spaces before inline comment*
- :bangbang: Line 491:47: **E262** *inline comment should start with '# '*
- :heavy_exclamation_mark: Line 496:44: **W291** *trailing whitespace*
- :heavy_exclamation_mark: Line 501:4: **W291** *trailing whitespace*
- :heavy_exclamation_mark: Line 502:21: **W291** *trailing whitespace*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 7ab189e5e387779747a961c62b56bd4223e131ae:

`cfme/common/__init__.py`:
- :bangbang: Line 392:74: **E261** *at least two spaces before inline comment*

`cfme/tests/infrastructure/test_host.py`:
- :bangbang: Line 491:50: **E261** *at least two spaces before inline comment*
- :bangbang: Line 491:51: **E262** *inline comment should start with '# '*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

prichard77 commented:
I updated test_infrastructure_hosts_navigation_after_download_from_compare but didn't include it in PRT because it is blocked by a BZ. It will fail anyway. I did remove the blocker and verify that the test fails where it should because of the BZ.

dajoRH commented:
Lint report for commit 14834920301f9e33242cec5e2a40994a0ef8bef6:

`cfme/common/__init__.py`:
- :bangbang: Line 392:74: **E261** *at least two spaces before inline comment*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

prichard77 commented:
Travis is failing because of cfme/tests/test_replication.py which I did not update and can't see the change it is referencing.

dajoRH commented:
Lint report for commit 95a59417f353b662251b5947e0edafff3a6e6525:

`cfme/infrastructure/virtual_machines.py`:
- :bangbang: Line 309:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 1366:5: **E265** *block comment should start with '# '*

`cfme/tests/infrastructure/test_providers.py`:
- :bangbang: Line 526:5: **E303** *too many blank lines (2)*
- :bangbang: Line 527:70: **E231** *missing whitespace after ','*
- :bangbang: Line 528:19: **E231** *missing whitespace after ':'*
- :bangbang: Line 539:5: **E303** *too many blank lines (2)*
- :heavy_exclamation_mark: Line 551:8: **W292** *no newline at end of file*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 7f820c97667d42ff2beee4d161576617cd533719:

`cfme/infrastructure/virtual_machines.py`:
- :bangbang: Line 309:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 1366:5: **E265** *block comment should start with '# '*

`cfme/tests/infrastructure/test_providers.py`:
- :bangbang: Line 526:5: **E303** *too many blank lines (2)*
- :bangbang: Line 527:70: **E231** *missing whitespace after ','*
- :bangbang: Line 528:19: **E231** *missing whitespace after ':'*
- :bangbang: Line 539:5: **E303** *too many blank lines (2)*
- :heavy_exclamation_mark: Line 551:8: **W292** *no newline at end of file*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 30604acffc1ccbdf3c6b9a92eaca27eefb99da80:

`cfme/infrastructure/virtual_machines.py`:
- :bangbang: Line 309:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 1366:5: **E265** *block comment should start with '# '*
- :bangbang: Line 1704:5: **E303** *too many blank lines (2)*
- :bangbang: Line 1713:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 1715:5: **E265** *block comment should start with '# '*
- :bangbang: Line 1724:45: **E261** *at least two spaces before inline comment*

`cfme/tests/infrastructure/test_providers.py`:
- :bangbang: Line 526:5: **E303** *too many blank lines (2)*
- :bangbang: Line 539:5: **E303** *too many blank lines (2)*
- :heavy_exclamation_mark: Line 551:8: **W292** *no newline at end of file*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit c058bcc1a2d2180053431af9354d4a53e83cbb65:

`cfme/infrastructure/virtual_machines.py`:
- :bangbang: Line 309:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 1366:5: **E265** *block comment should start with '# '*
- :bangbang: Line 1704:5: **E303** *too many blank lines (2)*
- :bangbang: Line 1713:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 1715:5: **E265** *block comment should start with '# '*
- :bangbang: Line 1724:45: **E261** *at least two spaces before inline comment*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

john-dupuy commented on cfme/infrastructure/host.py:
```diff
@@ -476,11 +476,17 @@ def cluster(self):
 
     ENTITY = Host
 
+    from cfme.common.host_views import ProviderHostsCompareView
+    from cfme.common.host_views import HostsCompareView
```
Why are these imports down here, avoiding import errors?

prichard77 replied:
Being lazy. I'll move to top of file.

john-dupuy commented on cfme/common/__init__.py:
```diff
@@ -365,6 +365,41 @@ def step(self, *args, **kwargs):
+        from all()), click on the compare dropdown item and verify the compare view is
+        displayed. It then returns the compare view.
+    """
+    DROPDOWN_TEXT = 'Compare Selected items'
+    NAV_STRING = 'All'
```
I think `COMPARE_VIEW` and `COMPARE_APP_VIEW` should have a default value or at least be explicitly declared in the mixin. Could you also put them in the docblock?

prichard77 replied:
Added to docblock and added to mixin class. I used a string default stating to overwrite. I could add the classes that HostsCollection (or another entity class) uses, but I'd need to import the classes and it seems a bit wonky because each class usually uses a unique value for this.

john-dupuy replied:
What about them defaulting to the generic class that you've defined below?

prichard77 replied:
Updated as suggested if I understood correctly.

prichard77 commented on cfme/common/__init__.py:
```diff
@@ -365,6 +365,41 @@ def step(self, *args, **kwargs):
+    NAV_STRING = 'All'
+
+    def compare_entities_col(self, provider, entities_list=None):
+        from cfme.utils.appliance.implementations.ui import navigate_to
```
Yes. But I could move it to the top of the file. I was debugging and it was easier to put it in the function.

john-dupuy replied:
Ah okay, if it's possible to move it to the top of the file, then I think we should.

prichard77 replied:
Done.

john-dupuy replied:
Still showing at L414 for me?

prichard77 commented on cfme/common/__init__.py:
```diff
@@ -365,6 +365,41 @@ def step(self, *args, **kwargs):
         self.prerequisite_view.toolbar.policy.item_select('Manage Policies')
 
 
+class ComparableMixin:
```
I initially had it this way, but something complained (I believe it was pre-commit hooks) so I removed '(object)' .

john-dupuy replied:
@prichard77 that's odd, we have many such usages of this in the framework. e.g. https://github.com/ManageIQ/integration_tests/blob/741be7d758b82c3d89c6d56a73ecf6c40b972c07/cfme/common/provider.py#L1143

prichard77 replied:
I believe pre-commit hooks may only be catching this when we have an occurrence in one of our commits. I am not 100% sure. I initially created the class with (object).

prichard77 replied:
I confirmed it is pyupgrade making this change. I added a dummy class to another PR (as to not mess this one up) and you can see the git diff here: https://travis-ci.com/github/ManageIQ/integration_tests/jobs/319657064

john-dupuy replied:
oh interesting! I'm not a fan of that, but oh well :(

dajoRH commented:
Lint report for commit 1edc5ce3060f5a831e1888a17397dce65b2cfa06:

`cfme/tests/infrastructure/test_vm_compare.py`:
- :red_circle: Line 1:1: **F401** *'fauxfactory' imported but unused*
- :red_circle: Line 3:1: **F401** *'widgetastic_patternfly.DropdownItemNotFound' imported but unused*
- :red_circle: Line 5:1: **F401** *'cfme.test_requirements' imported but unused*
- :red_circle: Line 8:1: **F401** *'cfme.infrastructure.provider.virtualcenter.VMwareProvider' imported but unused*
- :red_circle: Line 11:1: **F401** *'cfme.markers.env_markers.provider.providers' imported but unused*
- :red_circle: Line 13:1: **F401** *'cfme.utils.log.logger' imported but unused*
- :red_circle: Line 14:1: **F401** *'cfme.utils.providers.ProviderFilter' imported but unused*
- :bangbang: Line 38:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 47:1: **E265** *block comment should start with '# '*
- :bangbang: Line 52:37: **E127** *continuation line over-indented for visual indent*
- :bangbang: Line 64:66: **E127** *continuation line over-indented for visual indent*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Lint report for commit 9e82f9e04bbb3889c894e1859aed10dde4a561b8:

`cfme/tests/infrastructure/test_vm_compare.py`:
- :red_circle: Line 1:1: **F401** *'fauxfactory' imported but unused*
- :red_circle: Line 3:1: **F401** *'widgetastic_patternfly.DropdownItemNotFound' imported but unused*
- :red_circle: Line 5:1: **F401** *'cfme.test_requirements' imported but unused*
- :red_circle: Line 8:1: **F401** *'cfme.infrastructure.provider.virtualcenter.VMwareProvider' imported but unused*
- :red_circle: Line 11:1: **F401** *'cfme.markers.env_markers.provider.providers' imported but unused*
- :red_circle: Line 13:1: **F401** *'cfme.utils.log.logger' imported but unused*
- :red_circle: Line 14:1: **F401** *'cfme.utils.providers.ProviderFilter' imported but unused*
- :bangbang: Line 38:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 47:1: **E265** *block comment should start with '# '*
- :bangbang: Line 52:37: **E127** *continuation line over-indented for visual indent*
- :bangbang: Line 64:66: **E127** *continuation line over-indented for visual indent*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

john-dupuy commented on cfme/common/__init__.py:
```diff
@@ -365,6 +365,48 @@ def step(self, *args, **kwargs):
+            DROPDOWN_TEXT: (str) the text in the Configuration dropdown for comparing entities
+            NAV_STRING: (str) used as second arg in call to navigate_to()
+            COMPARE_VIEW: (view class) view class to create and return when filtering on provider
+            COMPARE_APP_VIEW: (view class) view class to create and return when no provider
```
What do mean by `no provider`?

john-dupuy replied:
Also I'm trying to think of a more descriptive name. Something like, `COMPARE_VIEW` and `COMPARE_VIEW_NO_PROVIDER`, what do you think?

prichard77 replied:
What about COMPARE_VIEW_PROVIDER and COMPARE_VIEW_ALL ? As you said, no provider is a bit misleading.

prichard77 commented on cfme/common/__init__.py:
```diff
@@ -365,6 +365,41 @@ def step(self, *args, **kwargs):
+class ComparableMixin:
+    """
+    Mixin for comparing entities. Should be added to Collection class.
+    Args:
+            provider: An InfraProvider instance
+            entities_list: A list of entities to compare
```
Moved as suggested.

prichard77 commented on cfme/common/host_views.py:
```diff
@@ -319,7 +319,23 @@ def is_displayed(self):
 
 
-class HostsCompareView(ComputeInfrastructureHostsView):
+# Should I move this to somewhere else?
+class CompareView(ComputeInfrastructureHostsView):
+    """generic class for compare views"""
```
Updated as suggested.

dajoRH commented:
Lint report for commit 04447fc195dc7c1c4348ac8e551e96cb15cd673d:

`cfme/tests/infrastructure/test_providers.py`:
- :bangbang: Line 514:37: **E127** *continuation line over-indented for visual indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

prichard77 commented on cfme/common/__init__.py:
```diff
@@ -365,6 +365,41 @@ def step(self, *args, **kwargs):
+            compare_entity_view = provider.create_view(self.COMPARE_VIEW)
+        else:
+            compare_entity_view = provider.create_view(self.COMPARE_APP_VIEW)
+        assert compare_entity_view.is_displayed
```
I was thinking it would be safer to assert in the mixin so the user wouldn't forget to do it and assume the compare went fine and also I was trying (overly so) to make my test small (one line). But I'll yield to your comment and assume this is the proper pattern used in the framework and we should follow moving forward. I only found 2 instances where assert view_obj.is_displayed was used outside of a test. I'll update as suggested.

john-dupuy commented on cfme/common/__init__.py:
```diff
@@ -365,6 +367,64 @@ def step(self, *args, **kwargs):
+            DROPDOWN_TEXT: (str) the text in the Configuration dropdown for comparing entities
+            NAV_STRING: (str) used as second arg in call to navigate_to()
+            COMPARE_VIEW_PROVIDER: (VIEW) view class to create and return when filtering on provider
+            COMPARE_APP_VIEW: (VIEW) view class to create and return when no provider filtering
```
To accomodate the name change
```suggestion
            COMPARE_VIEW_ALL: (VIEW) view class to create and return when no provider filtering
```

john-dupuy commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1339,9 +1354,22 @@ def genealogy(self):
+class InfraTemplateCollection(ComparableMixin, TemplateCollection):
     ENTITY = InfraTemplate
 
+    from cfme.common.provider_views import TemplatesCompareView
```
Can we move this to the top of the file?

john-dupuy commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -26,7 +26,9 @@
 from cfme.common import BaseLoggedInPage
+from cfme.common import ComparableMixin
 from cfme.common import TimelinesView
+from cfme.common.provider_views import TemplatesCompareView
```
Shouldn't `TemplatesCompareView` also inherit from your base `CompareView`? Or do you want to make those changes in a separate PR?

prichard77 replied:
Done.

john-dupuy commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1666,10 +1694,27 @@ def step(self, *args, **kwargs):
 
+    @property
+    def VIEW(self):  # noqa
+        from cfme.infrastructure.provider import ProviderTemplatesView
+        from cfme.infrastructure.provider import InfraProvider
```
Can these be moved to the top of the file?

prichard77 replied:
I moved ProviderTemplatesView to top of file, but I'll get import errors if I try to move InfraProvider. I am wondering if this is something I should look into with another PR or if I am missing a simple fix that could be added here.

john-dupuy commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1666,10 +1694,27 @@ def step(self, *args, **kwargs):
+        return TemplatesOnlyAllView
+
     def step(self, *args, **kwargs):
+        from cfme.infrastructure.provider import InfraProvider
```
Can we move this to the top of the file?

prichard77 replied:
As mentioned above, InfraProvider will cause import errors. I don't love having these imports but I feel it is a better check than just seeing if a parent exists. I have seen checks for just a parent in other places, but wanted to check for InfraProvider.

john-dupuy commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1666,10 +1694,27 @@ def step(self, *args, **kwargs):
+        from cfme.infrastructure.provider import InfraProvider
+        try:
+            if self.obj.parent.one_of(InfraProvider):
+                return ProviderTemplatesView  # ProviderTemplatesOnlyAllView?? What is this?
```
```suggestion
                return ProviderTemplatesView
```

prichard77 replied:
Removed.

john-dupuy commented on cfme/tests/infrastructure/test_providers.py:
```diff
@@ -520,14 +521,16 @@ def test_compare_provider_templates(appliance, setup_provider_min_templates, pro
-    view.toolbar.configuration.item_select('Compare Selected Templates', handle_alert=True)
-    compare_templates_view = provider.create_view(TemplatesCompareView)
-    assert compare_templates_view.is_displayed
-    assert compare_templates_view.verify_checked_items_compared(templateList,
-                                                                compare_templates_view)
```
This seems like a crucial part of the test case, why has it been removed?

prichard77 replied:
Thanks. Good catch. I left this off by mistake. Adding back.

dajoRH commented:
Lint report for commit b025471fb835a3e55f5e64a0e6dc6b68052a46a5:

`cfme/tests/infrastructure/test_providers.py`:
- :heavy_exclamation_mark: Line 539:76: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

mshriver commented on cfme/common/provider_views.py:
```diff
@@ -580,7 +581,7 @@ def before_fill(self, values):
+class TemplatesCompareView(CompareView):
     """Compare Templates page."""
     title = Text('.//div[@id="center_div" or @id="main-content"]//h1')
     comparison_table = Table(locator='//div[@id="compare-grid"]/table')
```
This should be in the common comparison view as well

prichard77 replied:
Updated as suggested.

mshriver commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -305,6 +308,19 @@ def is_displayed(self):
+class VmTemplatesCompareView(TemplatesCompareView):
+    """Compare Templates page."""
+    title = Text('.//div[@id="center_div" or @id="main-content"]//h1')
+    comparison_table = Table(locator='//div[@id="compare-grid"]/table')
```
Looks like the same widget, should be in the common view class.

prichard77 replied:
Updated as suggested.

john-dupuy commented on cfme/common/provider_views.py:
```diff
@@ -580,10 +580,8 @@ def before_fill(self, values):
 
 
-class TemplatesCompareView(InfraProvidersView):
+class TemplatesCompareView(CompareView):
```
:tada:

john-dupuy commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1339,9 +1353,21 @@ def genealogy(self):
+    NAV_STRING = 'TemplatesOnly'
+
+    @property
+    def name(self):
```
I would call this `provider_name`

prichard77 replied:
agreed. Updated.

prichard77 replied:
I could also add some conditionals in the code that references self.contect['object'] but that would seem to me to undo any clarity or readability we'd gain by renaming the property.

prichard77 replied:
I think I would update the property to provider_name and then add a variable name that would be `name = provider_name' .  That wouldn't clarify what .name is when calling it, but when looking at the class you would see that it's provider_name. Not sure how useful that would be.

prichard77 replied:
I could update the property to provider_name and then add a class variable name that would be `name = provider_name' . That wouldn't clarify what .name is when calling it, but when looking at the class you would see that it's provider_name. Not sure how useful that would be.

mshriver replied:
Whatever needs this name (view, nav class) should either base the attribute lookup on the type, or should be using `self.parent` to get the name attribute.

```
name = getattr(self.context['object'], 'name', None) or getattr(self.context['object'].parent, 'name', None)
```

prichard77 replied:
OK. I updated the view that needs this for another reason and now I should include support for removing the name property. In progress.

john-dupuy commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1339,9 +1353,21 @@ def genealogy(self):
+    NAV_STRING = 'TemplatesOnly'
+
+    @property
+    def name(self):
+        try:
+            return self.filters.get('provider').name
+        except Exception:
+            pass
```
```suggestion
    @property
    def name(self):
        provider = self.filters.get('provider')
        return provider.name if provider else None
```

prichard77 replied:
Updated as suggested.

mshriver commented on cfme/common/__init__.py:
```diff
@@ -365,6 +367,67 @@ def step(self, *args, **kwargs):
+    COMPARE_VIEW_PROVIDER = CompareView
+    COMPARE_VIEW_ALL = CompareView
+
+    def compare_entities_col(self, provider, entities_list=None):
```
nipick on the naming ... what's `_col`? collection?  I think `compare_entities` would be sufficient here.

prichard77 replied:
Updated as suggested.

mshriver commented on cfme/common/__init__.py:
```diff
@@ -365,6 +367,67 @@ def step(self, *args, **kwargs):
+
+    def compare_entities_col(self, provider, entities_list=None):
+        """
+        "Args:
```
extra quote mark here

prichard77 replied:
Updated.

prichard77 commented:
I also pushed an update to the name property in InfraTemplateCollection. I thought I had already pushed this.

dajoRH commented:
Lint report for commit 9c1efe4ad5caea6dfb7b6fed39ef74f960222fa6:

`cfme/infrastructure/virtual_machines.py`:
- :bangbang: Line 1370:5: **E303** *too many blank lines (2)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

mshriver commented on cfme/common/provider_views.py:
```diff
@@ -580,10 +580,8 @@ def before_fill(self, values):
 
 
-class TemplatesCompareView(InfraProvidersView):
+class TemplatesCompareView(CompareView):
```
The `verify_checked_items_compared`  method that was written for this class can be moved to the common view class. Can be in a separate PR.

mshriver replied:
That method needs some fixing anyway, it shouldn't be calling pytest.fail directly, and should be referencing `self` instead of taking a view instance as an arg.

prichard77 replied:
I made a note and will be fixing in forthcoming PR.

mshriver commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1666,10 +1690,26 @@ def step(self, *args, **kwargs):
+        try:
+            if self.obj.parent.one_of(InfraProvider):
+                return ProviderTemplatesView
+        except Exception:
```
What's the type of exception that you're trying to handle here? I think this is too much of a diaper.

prichard77 replied:
When parent is IPappliance this will cause ` AttributeError: 'IPAppliance' object has no attribute 'one_of'`

prichard77 replied:
I should probably just add one_of to IPAppliance ?

prichard77 replied:
As per our discussion, I'll use isinstance() and remove the try/except block.

prichard77 replied:
I removed usage of one_of and used isinstance() instead.

mshriver commented on cfme/tests/infrastructure/test_providers.py:
```diff
@@ -520,14 +521,19 @@ def test_compare_provider_templates(appliance, setup_provider_min_templates, pro
-    assert compare_templates_view.verify_checked_items_compared(templateList,
-                                                                compare_templates_view)
+    if templates_collection == 'provider':
+        provider.collections.templates = InfraTemplateCollection(provider,
```
These collections should not be directly instantiated, as with hosts it should be getting populated automatically.

prichard77 replied:
I am having a bit of trouble getting this to work without directly instantiating the collection for the provider case. I have it working for appliances. I need to determine 1) the correct way to get the collection to use in the call. 2) Any updates to the navigation that I need to do to support this. I believe I may be going in circles here. there are several "patterns" I am seeing in the framework, but I am likely using them incorrectly and don't really understand the underlying logic for many of these?

prichard77 replied:
No longer directly instantiated.

mshriver commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1666,10 +1690,26 @@ def step(self, *args, **kwargs):
+        from cfme.infrastructure.provider import InfraProvider
+        try:
+            if self.obj.parent.one_of(InfraProvider):
+                navigate_to(self.obj.parent, 'ProviderTemplates')
```
This is the prerequisite, not the step.

mshriver commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -452,30 +450,28 @@ def test_infrastructure_hosts_navigation_after_download(
-    compare_hosts_view = provider.create_view(ProviderHostsCompareView)
-    assert compare_hosts_view.is_displayed
+    if hosts_collection == 'provider':
+        h_coll = provider.hosts.all()[:num_hosts]
```
You can actually do this with `locals()[hosts_collection]` to get the instance that you're looking the collection up on.

`provider.collections.hosts` is equivalent to `provider.hosts`, so the lookup for either becomes the same.

```
locals()[hosts_collection].collections.hosts.all()
```

mshriver replied:
optional/FYI, I see the slicing of the result from `all()` is different, though its not clear why?

prichard77 replied:
OK. I see. Neat trick. I updated but used locals()[hosts_collection].collections.hosts to get just the collection, then made the list inside of the call with all()[:num_hosts] .

mshriver commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -452,30 +450,28 @@ def test_infrastructure_hosts_navigation_after_download(
+        compare_view = provider.hosts.compare_entities(provider, entities_list=h_coll)
+        assert compare_view.is_displayed
+    elif hosts_collection == 'appliance':
+        h_coll = appliance.collections.hosts.all()[-num_hosts:]
```
Why change the list slicing?

prichard77 replied:
Just to add another scenario to cover. We are always picking the first item. Thought we might should test using the last item returned.

prichard77 replied:
We don't do this any more. Same slices are used for iterations of the test.

prichard77 commented:
I did change a few things to get the compare templates tests to work so I'll be taking a closer look to make sure I didn't miss something. But I'd like to keep the review going on this PR.

prichard77 commented:
My last comment mentions that I'd be taking a closer look. I just wanted to clarify that this has been done for a bit and this is all ready for review.

mshriver commented on cfme/common/provider_views.py:
```diff
@@ -261,8 +261,12 @@ class ProviderTemplatesView(ProviderVmsTemplatesView):
         title = VersionPicker({
-            Version.lowest(): '{name} (All VM Templates)'.format(name=self.context['object'].name),
-            '5.10': '{name} (All VM Templates and Images)'.format(name=self.context['object'].name),
+            Version.lowest(): '{name} (All VM Templates)'.format(
```
Optional, consider cleaning up in follow up - the only difference here is part of the string, so you can separate out the duplicated logic of picking the object name or the object parent name.

```python
expected_name = getattr(self.context['object'], 'name', None) or getattr(context['object'].parent, 'name', None)
title = VersionPicker({Version.lowest(): f'{expected_name} (All VM Templates)', '5.10': f'{expected_name} (All VM Templates and Images)'})
```

prichard77 replied:
Sure, I agree. I was simply updating what was here with a small change without rewriting it. I will name a note to update this and also other instances of it that exist in the code (There are several).

mshriver commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1666,17 +1688,34 @@ def step(self, *args, **kwargs):
+
+    def prerequisite(self):
+        from cfme.infrastructure.provider import InfraProvider
+        if isinstance(self.obj.parent, InfraProvider):
```
I'm a bit concerned about this import/comparison, I think it could be done without actually importing these classes.

Can you check and verify that this could be achieved by comparing like:

```
if self.obj.parent is self.obj.provider:
    nav_inst = self.obj.parent
    nav_dest = 'Details'
elif self.obj.parent is self.obj.appliance:
    nav_inst = self.obj
    nav_dest = 'All'
# things blow up if nav_dest or nav_inst isn't set, maybe look for them in locals() or init them
return navigate_to(nav_inst, nav_dest)

```

mshriver replied:
At a minimum, I'd want the import of IPAppliance moved to this local scope, just like InfraProvider, to avoid circular import potential.

prichard77 replied:
I implemented something similar that removes the need to import.

prichard77 replied:
I implemented NAV_STRING as a property in the Collection classes and had it pass the appropriate string based upon the existence of  a provider filter. Removed this logic from the navigation step class.

mshriver commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1666,17 +1688,34 @@ def step(self, *args, **kwargs):
 class TemplatesAll(CFMENavigateStep):
-    VIEW = TemplatesOnlyAllView
-    prerequisite = NavigateToSibling('All')
+
```
Check out how `InfraTemplateCollection.all` method is changing its navigation destination based on the filter. The changes you're making to this navigation class are actually crossing over with the existing `ProviderTemplates` destination that's registered against the provider instance.

ComparableMixin could pick its `navigate_to` call based on the parent/filter, instead of trying to push that logic down here into the navigation destination.

prichard77 replied:
Just to be clear, this is to be used instead of 
```
if self.obj.parent is self.obj.appliance:
            nav_inst = self.obj
            nav_dest = 'All'
        else:
            nav_inst = self.obj.parent
            nav_dest = 'Details'
        return navigate_to(nav_inst, nav_dest)``` 
Makes sense, and we can leave the nav destination as is was.

prichard77 replied:
I implemented NAV_STRING as a property in the Collection classes and had it pass the appropriate string based upon the existence of a provider filter.

mshriver commented on cfme/tests/infrastructure/test_providers.py:
```diff
@@ -514,14 +514,13 @@ def test_compare_provider_templates(appliance, setup_provider_min_templates, pro
+        entities_list=t_coll)
+    assert compare_view.is_displayed
+
+    t_list = []
```
Optional, consider using list comprehension here. `templates = [t.name for t in t_coll]`

prichard77 replied:
Of course.

prichard77 replied:
FYI. I still have a rewrite of verify_checked_items_compared on my todo list.

mshriver commented on cfme/tests/infrastructure/test_host.py:
```diff
@@ -493,23 +488,14 @@ def test_infrastructure_hosts_navigation_after_download_from_compare(
     hosts_view.navigation.select("Compute")
-    compare_hosts_view = provider.create_view(HostsCompareView)
-    assert compare_hosts_view.is_displayed
+    assert hosts_view.is_displayed
```
Optional: 
Can we include the call to `assert compare_view.verify_checked_items_compared(t_list, compare_view)` here as well for more thorough coverage?

mshriver replied:
Nevermind, I see the method is currently only available for Templates View.

prichard77 commented:
I have implemented all suggested changes except for the optional VersionPicker update. I'll put this update in another PR if I don't have time to get it in and test it before this is reviewed again and merged.

prichard77 commented:
I made the suggested updates and got everything working with one COMPARE_VIEW and the NAV_STRING logic moved into the Collection instead of the navigation step class. I also went ahead and updated VirsionPicker. I'll be standing by most of the day Thursday for any comments.

prichard77 commented:
Fixed a bug in HostCollection.

prichard77 commented:
I didn't really want to re-request a review from Mike and John. I was poking around in the UI and clicked the wrong icon. They are both already aware, have been commenting.

mshriver commented on cfme/infrastructure/virtual_machines.py:
```diff
@@ -1670,13 +1696,17 @@ class TemplatesAll(CFMENavigateStep):
+                raise DestinationNotFound("the destination isn't found")
         else:
-            raise DestinationNotFound("the destination isn't found")
+            self.prerequisite_view.entities.summary("Relationships").click_at("Templates")
```
Have you tested this navigation path? The prerequisite view here is InfraTemplateCollection's 'All' destination, which won't have a relationships summary table.

mshriver replied:
What I was trying to point out in previous comments is that this navigation destination already exists as `ProviderTemplates` for the provider instance. `InfraTemplateCollection.all` is using it.

https://github.com/ManageIQ/integration_tests/blob/master/cfme/infrastructure/provider/__init__.py#L446

The navigation instance and destination need to be picked to be able to use this, and I think that can be done in Comparable.

prichard77 replied:
I am looking at this now and testing, but I believe I handled this before I get to TemplatesAll. I should have deleted the extra code meant for providers , but I don't think I'd ever get here with a provider, as I changed InfraTemplateCollection to pass NAV_STRING and COMPARE_VIEW based upon having a provider filter or not. I am going to go though and test everything just to make sure and will fix anything I find. Thanks.

prichard77 replied:
Yeah. I meant to remove the code I added to TemplatesAll. It has been removed now, but I have a bug in the navigation to work out. I have removed the logic for determining provider vs. appliance out of navigation step and view classes. Should have the bug fixed shortly. I am testing thoroughly.

prichard77 commented:
I figured out what was throwing me off. The code in InfraTemplateCollection would have never worked for providers but seemed to be coded to support it. It also explians why I was having issues with creating the provider collections early on. There still may be a small item or two left to cleanup or optimize, but I have tested hosts and template compare scenarios. Might also want to look into using the api to create the collections.

prichard77 commented:
I also believe I could just register ProviderTemplates with InfraTemplatesCollection in navigation and remove the try/except, but I haven't looked into it closely. I'll give it a try in the morning.

mshriver commented on cfme/common/provider_views.py:
```diff
@@ -260,10 +260,13 @@ class ProviderTemplatesView(ProviderVmsTemplatesView):
+            self.context['object'].parent, 'name', None)
+        title = VersionPicker(
+            {Version.lowest(): f'{expected_name} (All VM Templates)',
+            '5.10': f'{expected_name} (All VM Templates and Images)'
```
Didn't catch this earlier, but this no longer needs to be version picked. Lowest supported version of CFME in integration_tests is 5.10. Its okay to leave it in, but note its not necessary to include the 5.9 string here via Version.lowest.

mshriver commented on cfme/common/__init__.py:
```diff
@@ -365,6 +367,65 @@ def step(self, *args, **kwargs):
+    NAV_STRING = 'All'
+    COMPARE_VIEW = CompareView
+
+    def compare_entities(self, provider, entities_list=None):
```
Please remove provider as required arg here - its only used for creating the view, and you can call `create_view` on `self` here, as its provided by NavigatableMixin on the collections.
